###Data Preperation H3K27ac
# Input files merged replicate bam files using bedtools merge.

#Broad-Peak q-value
macs2 callpeak -t /mnt/RawChIP_Histon/Merged/ChIP_Seq_H3K27ac_merged_cle_dd.bam -c /mnt/RawChIP_Histon/Merged/ChIP_Seq_control_Input_merged_cle_dd.bam -f BAMPE --bdg --broad --broad-cutoff 0.01 -q 0.01 -g 2913022398 --outdir /mnt/RawChIP_Histon/MACS2 -n ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_broadq0.01macs2

#Narrow-Peak q-value
macs2 callpeak -t /mnt/RawChIP_Histon/Merged/ChIP_Seq_H3K27ac_merged_cle_dd.bam -c /mnt/RawChIP_Histon/Merged/ChIP_Seq_control_Input_merged_cle_dd.bam -f BAMPE --bdg -q 0.01 -g 2913022398 --outdir /mnt/RawChIP_Histon/MACS2 -n ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_q0.01macs2


#######################
###ROSE################
#######################


Setup Rose

Download frot github
suod git clone https://github.com/stjude/ROSE.git

sudo wget https://bitbucket.org/young_computation/rose/src/master/

PATHTO=/mnt/ROSE2/ROSE
PYTHONPATH=$PATHTO/lib
sudo export PYTHONPATH
sudo export PATH=$PATH:$PATHTO/bin


#Generate Fold Enrichment traces

macs2 bdgcmp -t /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_broadq0.01macs2_treat_pileup.bdg -c /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_broadq0.01macs2_control_lambda.bdg -m FE --o-prefix /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_broadq0.01macs2

Fold Enrichment traces were sorted using igvtools sort <input>_FE.bdg <input>_FE.sorted.bdg hg38
TDF files were generated for IGV display using igvtools toTDF -z 10 <input>_FE.sorted.bdg <input>_FE.tdf hg38


#Convert broad or narrow peak into bed

bedtools convert -i /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_q0.01macs2 -o bed -header > /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_q0.01macs2.bed
bedtools convert -i /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_broadq0.01macs2 -o bed -header > /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_q0.01macs2.bed

bedtools merge -i /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_broadq0.01macs2_peaks.broadPeak.bed > /mnt/RawChIP_Histon/MACS2/stitched_enhancers_H3K27ac_broadpeaks.bed

bedtools merge -i /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_q0.01macs2_peaks.narrowPeak.bed > /mnt/RawChIP_Histon/MACS2/stitched_enhancers_H3K27ac_narrowpeaks.bed

#ROSE was used to call superenhancers from H3K27ac ChIP data using the following command: 

python ROSE_main.py -g HG38 -i /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_q0.01macs2_peaks.narrowPeak.bed -r /mnt/RawChIP_Histon/Merged/ChIP_Seq_H3K27ac_merged_cle_dd.bam -o /mnt/RawChIP_Histon/ROSE
python ROSE_main.py -g HG38 -i /mnt/RawChIP_Histon/MACS2/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_broadq0.01macs2_peaks.broadPeak.bed -r /mnt/RawChIP_Histon/Merged/ChIP_Seq_H3K27ac_merged_cle_dd.bam -o /mnt/RawChIP_Histon/ROSE 


python ROSE_main.py -g HG38 -i /mnt/ROSE/Input/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_broadq0.01macs2_peaks.broadPeak.bed -r /mnt/ROSE/Input/ChIP_Seq_H3K27ac_merged_cle_dd.bam -o /mnt/ROSE/Input


python ROSE_main.py -g HG38 -i /mnt/RawChIP_Histon/MACS2/HAL01_H3K27ac_merged_cle_sort_dd_p9bmacs2_peaks.broadPeak.bed -r /mnt/RawChIP_Histon/Merged/ChIP_Seq_H3K27ac_merged_cle_dd.bam -o /mnt/RawChIP_Histon/ROSE -s 6000 -t 2500 -c /mnt/RawChIP_Histon/Merged/ChIP_Seq_control_Input_merged_cle_dd.bam


python /mnt/ROSE/bin/ROSE_main.py -g HG38 -i /mnt/ROSE/Input/data/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_broadq0.01macs2_peaks.broadPeak.bed -r /mnt/ROSE/Input/data/ChIP_Seq_H3K27ac_merged_cle_dd.bam  -o HAL01H3K27acbroad

python /mnt/ROSE/bin/ROSE_main.py -g HG38 -i /mnt/ROSE/Input/data/ChIP_Seq_HAL01_H3K27ac_merged_cle_dd_q0.01macs2_peaks.narrowPeak.bed -r /mnt/ROSE/Input/data/ChIP_Seq_H3K27ac_merged_cle_dd.bam  -o HAL01H3K27acnarrow


 Optional parameters:

    STITCHING_DISTANCE: maximum distance between two regions that will be stitched together (Default: 12.5kb)
    TSS_EXCLUSION_ZONE_SIZE: exclude regions contained within +/- this distance from TSS in order to account for promoter biases (Default: 0; recommended if used: 2500). If this value is 0, will not look for a gene file.
    CONTROL_BAM: .bam file to be used as a control. Subtracted from the density of the RANKING_BAM. I.e. Whole cell extract reads.


Supplementary_files_format_and_content: .narrowPeak files: bed6+4 containing peak calls from macs2
Supplementary_files_format_and_content: .broadPeak files: bed6+4 containing broad peak calls from macs2
Supplementary_files_format_and_content: .tdf files: IGV compressed files containing fold enrichment tracks. Generated from MACS2 output using IGVtools.
Supplementary_files_format_and_content: .bed files - union peak call regions from MACS2 or superenhancer calls generated by ROSE

